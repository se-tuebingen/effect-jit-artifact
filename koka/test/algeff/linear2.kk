
effect state<s> {
  ctl get() : s
  ctl put(x : s) : ()
}

fun state(init,action) {
  var s := init
  with {
    fun get(){ s }
    ctl put(t){ s := t; val x = resume(()); s := t; 20 + x + resume(()) } 
  }
  action()
}


fun foo() {
  val i = get()
  put(i+1)
  put(i+2)
  get()
}

fun main() {
  state(0,foo).println  // 68
}