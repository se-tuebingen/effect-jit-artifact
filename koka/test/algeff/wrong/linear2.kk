module wrong/linear2

effect amb {
  ctl flip() : bool
}

val amb = handler {
  return(x)  -> [x]
  ctl flip() -> resume(True) ++ resume(False)
}

linear effect state<s> {
  fun get() : s
  fun put(x : s) : ()
}

fun state(init,action) 
  var s := init
  handle(action)
    fun get()  { s }
    fun put(t) { if flip() then s := t; () }

fun foo() {
  val i = get()
  put(i+1)
  put(i+2)
  get()
}

pub fun main() {
  amb({state(0,foo)}).showList(show).print  // 42
}