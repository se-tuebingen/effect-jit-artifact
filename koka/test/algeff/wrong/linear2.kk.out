test/algeff/wrong/linear2@kk(19, 3): type error: handler for state needs to be linear but uses a non-linear effect: amb
 context: handle(action)
 fun get() { s }
 fun put(t) { if flip() then s := t; () }
 term : handle
 hint : ensure only linear effects are used in a handler